File: 20230125_Wednesday/20230125.rb
# Leetcode: 2359. Find Closest Node to Given Two Nodes.
# https://leetcode.com/problems/find-closest-node-to-given-two-nodes/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 296 ms, faster than 100.00% of Ruby online submissions for Find Closest Node to Given Two Nodes.
# Memory Usage: 231.5 MB, less than 100.00% of Ruby online submissions for Find Closest Node to Given Two Nodes.
# 2023.01.25 Daily Challenge.
# @param {Integer[]} edges
# @param {Integer} node1
# @param {Integer} node2
# @return {Integer}
def closest_meeting_node(edges, node1, node2)
  v1 = [false] * edges.size
  d1 = [0] * edges.size
  q = [[node1,0]]
  v1[node1] = true
  while ! q.empty?
    x,l = q.shift
    d1[x] = l
    y = edges[x]
    if -1 != y && (!v1[y])
      q.push([y,l+1])
      v1[y] = true
    end
  end
  v2 = [false] * edges.size
  d2 = [0] * edges.size
  q = [[node2,0]]
  v2[node2] = true
  while ! q.empty?
    x,l = q.shift
    d2[x] = l
    y = edges[x]
    if -1 != y  && (!v2[y])
      q.push([y,l+1])
      v2[y]= true
    end
  end
  answer = -1
  min = edges.size * 10
  edges.size.times do |i|
    if v1[i] && v2[i]
      tmp = [d1[i],d2[i]].max # Error fixed here! Was min.
      if tmp < min
        min = tmp
        answer = i
      end
    end
  end
  return answer
end

File: 20230125_Wednesday/20230125.go
// Leetcode: 2359. Find Closest Node to Given Two Nodes.
// https://leetcode.com/problems/find-closest-node-to-given-two-nodes/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 166 ms, faster than 63.64% of Go online submissions for Find Closest Node to Given Two Nodes.
// Memory Usage: 9.6 MB, less than 81.82% of Go online submissions for Find Closest Node to Given Two Nodes.
// 2023.01.25 Daily Challenge.
import "fmt"
var p = fmt.Println
func closestMeetingNode(edges []int, node1 int, node2 int) int {
  n := len(edges)
  d1 := make([]int,n)
  qv := make([]int,1) // Vertex.
  ql := make([]int,1) // Level.
  qv[0],ql[0] = node1, 0
  v1 := make([]bool,n)
  v1[node1] = true
  for 0 != len(qv) {
    x := qv[0]
    l := ql[0]
    y := edges[x]
    qv = qv[1:len(qv)]
    ql = ql[1:len(ql)]
    d1[x] = l
    if -1 != y && (!v1[y]) {
      ql = append(ql,l+1)
      qv = append(qv,y)
      v1[y] = true
    }
  }
  v2,d2 := make([]bool,n),make([]int,n)
  v2[node2] = true
  qv,ql = make([]int,1),make([]int,1)
  qv[0],ql[0] = node2, 0
  for 0 != len(qv) {
    x, l := qv[0], ql[0]
    y := edges[x]
    qv = qv[1:len(qv)]
    ql = ql[1:len(ql)]
    d2[x] = l
    if -1 != y && (!v2[y]){
      ql = append(ql, l+1)
      qv = append(qv,y)
      v2[y] = true
    }
  }
  min := n*10
  answer := -1
  for i := 0; i < n ; i += 1 {
    if v1[i] && v2[i] {
      tmp := max(d1[i],d2[i])
      if tmp < min {
        answer = i
        min = tmp
      }
    }
  }
  return answer
}

func max(a, b int) int {
  if a > b {
    return a 
  } else {
    return b
  }
}
